### PR TITLE
GEODE-2936: Refactoring OrderByComparator

### PR DESCRIPTION
**Pull request by YehEmily**
[View the JIRA ticket here.](https://issues.apache.org/jira/browse/GEODE-2936)

`OrderByComparator.java` contained a lot of redundant and confusing code, as well as somewhat misleading comments (for example, the `compare` method was supposed to throw a `ClassCastException` if the parameters given could not be compared using this comparator, but it didn't. This has been fixed). Code that is shared by both the `compare` and `evaluateSortCriteria` methods has been moved to helper methods to improve readability and reduce redundancy.

_**Precheckin status**: successful after second round of tests (6/27/17)_

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java [hunk 1]
```diff
@@ -45,78 +44,50 @@ public OrderByComparator(List<CompiledSortCriterion> orderByAttrs, ObjectType ob
...
       array = new Object[orderByAttrs.size()];
-      Iterator orderiter = orderByAttrs.iterator();
+      Iterator orderIter = orderByAttrs.iterator();
       int i = 0;
-      while (orderiter.hasNext()) {
-        csc = (CompiledSortCriterion) orderiter.next();
-        Object[] arr = new Object[2];
-        arr[0] = csc.evaluate(value, context);
-        arr[1] = Boolean.valueOf(csc.getCriterion());
+      while (orderIter.hasNext()) {
```
Comment (PurelyApplied) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java]: This loop and the one below could read better with a "for each" loop.
```
for (CompiledSortCriterion csc: orderByAttrs){
```

Good refactoring overall.
+1 with or without the loop edit.

Comment (YehEmily) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java]: Fixed, and updating the PR now - thank you!

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java [hunk 2]
```diff
@@ -228,4 +139,55 @@ void addEvaluatedSortCriteria(Object row, ExecutionContext context)
...
+
+  private int compareTwoObjects(Object obj1, Object obj2) {
+    if (obj1 instanceof Number && obj2 instanceof Number) {
+      return compareTwoNumbers(obj1, obj2);
+    } else {
+      return compareTwoStrings(obj1, obj2);
+    }
+  }
+
+  private int compareTwoNumbers(Object obj1, Object obj2) {
```
Comment (jaredjstewart) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java]: I see it was not introduced by your changes, but the subtraction on 177 is not a safe way to compare doubles as it can result in under/overflow.  For example, consider the result of `compareTwoNumbers( Double.MAX_VALUE, -1d)`.  Here's a safer way to do it:

```
  private int compareTwoNumbers(Object obj1, Object obj2) {
    Number number1 = (Number) obj1;
    Number number2 = (Number) obj2 ;
    return Double.compare(number1.doubleValue(), number2.doubleValue());
  }
```

Comment (YehEmily) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/OrderByComparator.java]: Oh no, that's very true! I'll fix this and update the PR!

**Comment by YehEmily** [comment]
@jhuynh1 Thanks for the suggestion! I updated the tests in `OrderByComparatorJUnitTest.java` to cover all the cases in `OrderByComparator.java` and updated the PR.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java [hunk 3]
```diff
@@ -173,36 +157,58 @@ public void testUnsupportedOrderByForPR() throws Exception {
...
-
       } catch (QueryInvalidException qe) {
         qe.printStackTrace();
         fail(qe.toString());
       }
     }
   }
 
+  /**
+   * Tests three cases that were not originally covered by the original tests in this class. (To
```
Comment (pdxrunner) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: This comment probably doesn't need to be a javadoc, and the note about coverage isn't needed. I'd make this comment say something like, "The following tests cover edge cases in OrderByComparator".

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java [hunk 4]
```diff
@@ -173,36 +157,58 @@ public void testUnsupportedOrderByForPR() throws Exception {
...
+   * confirm that these tests cover all cases in OrderByComparator, run them with coverage.)
+   */
+  @Test
+  public void testCompareMethodEdgeCases() throws Exception {
```
Comment (pdxrunner) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: To adhere to the guideline that each test should only check thing, I would break this up into separate tests.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java (hunk 3 above)
Comment (YehEmily) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: All good points - I've fixed this and updated the PR!

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java (hunk 4 above)
Comment (YehEmily) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: Thanks for the feedback! I've updated the PR!

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java [hunk 5]
```diff
@@ -173,36 +157,54 @@ public void testUnsupportedOrderByForPR() throws Exception {
...
         qe.printStackTrace();
         fail(qe.toString());
       }
     }
   }
 
+  // The following tests cover edge cases in OrderByComparator.
+  @Test
+  public void testCompareTwoNulls() throws Exception {
+    assert (createComparator().compare(null, null) == 0);
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: "assert" should be used in product code instead of tests. These are the optional assertions that are enabled when running a Java product with -ea (enable assertions).

You should use assertThat instead:
```java
assertThat(createComparator().compare(null, null).isEqualTo(0);
```

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java [hunk 6]
```diff
@@ -173,36 +157,54 @@ public void testUnsupportedOrderByForPR() throws Exception {
...
+  }
+
+  @Test
+  public void testCompareTwoObjectArrays() throws Exception {
+    String[] arrString1 = {"elephant"};
+    String[] arrString2 = {"elephants"};
+    assert (createComparator().compare(arrString2, arrString1) == 1);
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: Same as previous comment. Please use assertThat.

File: geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java (hunk 5 above)
Comment (YehEmily) [inline geode-core/src/test/java/org/apache/geode/cache/query/internal/OrderByComparatorJUnitTest.java]: Thanks for catching this! I've fixed it and updated the PR.
